### PR TITLE
It's the `final` countdown

### PR DESCRIPTION
**Pull request by m-kuhn**
Teach sip the meaning of final and make use of it in QgsVectorLayer (which is not supposed to be subclassed)

### CODE REVIEW

**Comment by luipir** [comment]
no please... the Europe noooooo

**Comment by m-kuhn** [comment]
@3nids do you know by any chance if there's a minimum sip requirement? The `final` keyword is supported starting from [4.19](https://www.riverbankcomputing.com/news/sip-419).

**Comment by 3nids** [comment]
we're far away from 4.19.....
I think we're even at 4.17.

**Comment by m-kuhn** [comment]
So... we do a `SIP_FINAL`?

**Comment by m-kuhn** [comment]
Oh wait... we can only do that if we generate .sip files at build time where we know the sip version.

**Comment by 3nids** [comment]
what you have to do is...
1. Only declare SIP_FINAL
2. In sipfiy, make it write ${SIP_FINAL}
3. In python/CmakeLists (around line 162), add a new test on SIP version and do:
```
IF(${SIP_VERSION_STR} VERSION_GREATER 4.19.XXXX)
  SET(SIP_FINAL "final")
ELSE(${SIP_VERSION_STR} VERSION_GREATER 4.19.XXXX)
  SET(SIP_FINAL "override")
END(${SIP_VERSION_STR} VERSION_GREATER 4.19.XXXX)
 ```

File: src/core/qgis_sip.h [hunk 1]
```diff
@@ -197,6 +197,16 @@
  */
 #define SIP_DOC_TEMPLATE
 
+/*
+ * Sip supports the final keyword since version 4.19.0, earlier than that
+ * we will have build issues because it tries to override final methods.
+ */
+#if SIP_VERSION >= 0x041300
+#define SIP_FINAL final
```
Comment (nyalldawson) [inline src/core/qgis_sip.h]: just wondering -- could we avoid using SIP_FINAL in the cpp headers and just use "final" instead?

Comment (m-kuhn) [reply src/core/qgis_sip.h]: And have a macro (in qgis.h?) that redefines final to override for older sip versions?

Comment (nyalldawson) [reply src/core/qgis_sip.h]: couldn't we do the replacement in sipify?

Comment (m-kuhn) [reply src/core/qgis_sip.h]: Cpp headers also need to be different.
If you have a `final` in a header and use it with sip 4.18, sip will override this method and the compiler will complain.

Comment (nyalldawson) [reply src/core/qgis_sip.h]: Ok. I'm late to the discussion :p

Hmm messy either way. I'd prefer a more generic macro if nothing else, even "FINAL" or "QGS_FINAL"

Comment (m-kuhn) [reply src/core/qgis_sip.h]: I was thinking the same.
And it should actually be defined by qgis.h and redefined by qgis_sip.h for older sip versions. But the effect will be, that someone with an older sip version will have to do an include order cleanup ;P

Comment (m-kuhn) [reply src/core/qgis_sip.h]: I guess I was able to avoid the todo

**Comment by m-kuhn** [comment]
Thanks for the help @3nids !

**Review by 3nids — APPROVED** [review]
cool!

File: src/core/qgis_sip.h [hunk 2]
```diff
@@ -197,6 +197,17 @@
...
 
+/*
+ * Sip supports the final keyword since version 4.19.0, earlier than that
+ * we will have build issues because it tries to override final methods.
+ */
+#if SIP_VERSION < 0x041300
+#if defined FINAL
+#undef FINAL
+#endif
+#define FINAL override
```
Comment (3nids) [inline src/core/qgis_sip.h]: missing #else

Comment (m-kuhn) [reply src/core/qgis_sip.h]: No else required, that's defined by qgis.h

**Comment by m-kuhn** [comment]
https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=2ahUKEwiw44StjMfdAhVGposKHW2OCHIQyCkwAHoECAYQBQ&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3D9jK-NcRmVcw&usg=AOvVaw1W9hAKdJJf7LMOqKIussAz

**Comment by stev-0** [comment]
The `override` keyword is giving me syntax errors which go away when it is replaced with nothing as before and the the project can be built (on SIP < 4.19).

**Comment by 3nids** [comment]
@stev-0 should be fixed in https://github.com/qgis/QGIS/commit/3ceecbf8063e97ba721442b6ab15f159a352281b

**Comment by stev-0** [comment]
I spoke too soon - I still get the following build errors when using SIP < 4.17. Sorry, this probably isn't the best place for this, but pretty sure it's related to this PR, as I built it fine with the commit before. Any ideas? I have tried a few things a little blindly but can't get it to build without errors.

    /home/stephen/dev/cpp/QGIS/build-master/python/core/sip_corepart1.cpp:232499:22: error: invalid covariant return type for ‘virtual QgsDataProvider* sipQgsAuxiliaryLayer::dataProvider()’
    In file included from /home/stephen/dev/cpp/QGIS/src/core/qgsvectorlayerref.h:19:0,
                 from /home/stephen/dev/cpp/QGIS/src/core/layout/qgsabstractreportsection.h:22,
                 from /home/stephen/dev/cpp/QGIS/src/core/layout/qgsreport.h:20,
                 from /home/stephen/dev/cpp/QGIS/build-master/python/core/auto_generated/layout/qgsmasterlayoutinterface.sip:13,
                 from /home/stephen/dev/cpp/QGIS/build-master/python/core/sip_corepart1.cpp:7:
    /home/stephen/dev/cpp/QGIS/src/core/qgsvectorlayer.h:487:28: error:   overriding ‘virtual QgsVectorDataProvider* QgsVectorLayer::dataProvider()’
     QgsVectorDataProvider *dataProvider();
                            ^
    /home/stephen/dev/cpp/QGIS/build-master/python/core/sip_corepart1.cpp:236190:22: error: invalid covariant return type for ‘virtual QgsDataProvider* sipQgsVectorLayer::dataProvider()’
     memset(sipPyMethods, 0, sizeof (sipPyMethods));

**Comment by m-kuhn** [comment]
Can you try that one?
https://github.com/qgis/QGIS/pull/7971

**Comment by stev-0** [comment]
Yup, looks good.

**Pull request merged**